src/Transport/Receiving/MessageExtensions.cs

```diff
@@ -0,0 +1,55 @@
+﻿namespace NServiceBus.Transport.AzureServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+    using Configuration;
+    using Microsoft.Azure.ServiceBus;
+    using Microsoft.Azure.ServiceBus.InteropExtensions;
+
+    static class MessageExtensions
+    {
+        public static Dictionary<string, string> GetNServiceBusHeaders(this Message message)
+        {
+            var headers = new Dictionary<string, string>(message.UserProperties.Count);
```

Comment:
It would be so nice if we could buffer dictionaries. Probably something for a next version but in theory we know how many concurrent messages we can handle so we could allocate a buffer of dictionaries and then reuse and clear them

Reply:
I might be off, but how much of an overhead we're talking about to justify the added complexity?

Reply:
I'd definitely want to see some benchmarks before making this sort of change.

Reply:
Feel free to submit a PR when you have good evidence that performance gain justifies the effort. Until then, marking as nice-to-have.